File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/adapterdelegates/diffutil/OffsetUpdateCallback.java
package com.github.klee0kai.hummus.adapterdelegates.diffutil;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

/**
 * dispatch updates with offset
 */
public class OffsetUpdateCallback implements IListUpdateCallbackExt {

    private final int offset;
    private RecyclerView.Adapter adapter;

    public OffsetUpdateCallback(RecyclerView.Adapter adapter, int offset) {
        this.offset = offset;
        this.adapter = adapter;
    }


    public OffsetUpdateCallback(RecyclerView.Adapter adapter) {
        this(adapter, 0);
    }


    @Override
    public void onInserted(int position, int count) {
        adapter.notifyItemRangeInserted(position + offset, count);
    }

    @Override
    public void onRemoved(int position, int count) {
        adapter.notifyItemRangeRemoved(position + offset, count);

    }

    @Override
    public void onMoved(int fromPosition, int toPosition) {
        adapter.notifyItemMoved(fromPosition + offset, toPosition + offset);
    }

    @Override
    public void onChanged(int position, int count, @Nullable Object payload) {
        adapter.notifyItemRangeChanged(position + offset, count, payload);
    }

    @Override
    public void updateAll() {
        adapter.notifyDataSetChanged();
    }
}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/adapterdelegates/diffutil/OffsetUpdateCallback.kt
package com.github.klee0kai.hummus.adapterdelegates.diffutil

import androidx.recyclerview.widget.RecyclerView

/**
 * dispatch updates with offset
 */
open class OffsetUpdateCallback(
    protected open val adapter: RecyclerView.Adapter<*>,
    protected open val offset: Int = 0
) : IListUpdateCallbackExt {

    override fun onInserted(position: Int, count: Int) {
        adapter.notifyItemRangeInserted(position + offset, count)
    }

    override fun onRemoved(position: Int, count: Int) {
        adapter.notifyItemRangeRemoved(position + offset, count)
    }

    override fun onMoved(fromPosition: Int, toPosition: Int) {
        adapter.notifyItemMoved(fromPosition + offset, toPosition + offset)
    }

    override fun onChanged(position: Int, count: Int, payload: Any?) {
        adapter.notifyItemRangeChanged(position + offset, count, payload)
    }

    override fun updateAll() {
        adapter.notifyDataSetChanged()
    }
}